File: ruby/lib/br_validator/ie/am.rb
# frozen_string_literal: true

require_relative "../shared"
require_relative "mod11"

module BrValidator
  module Ie
    # AM: verified against the official SEFAZ-AM "Roteiro de Crítica da
    # Inscrição Estadual" (sintegra.gov.br/Cad_Estados/cad_AM.html).
    # Format: 99.999.999-9 (8 digits + 1 check digit). Unlike most other
    # states, when the weighted sum itself is below 11 the digit is 11
    # minus the sum directly (skipping the modulo step).
    module Am
      module_function

      ALLOWED_CHARS = /\A[\d.\-\s]+\z/.freeze
      WEIGHTS = [9, 8, 7, 6, 5, 4, 3, 2].freeze
      private_constant :ALLOWED_CHARS, :WEIGHTS

      def normalize(value)
        Shared.remove_non_digits(value)
      end

      def calculate_check_digit(v)
        sum = Mod11.weighted_sum(v, WEIGHTS)

        return 11 - sum if sum < 11

        Mod11.mod11_check_digit(sum)
      end
      private_class_method :calculate_check_digit

      def is_valid?(value)
        return false unless value.match?(ALLOWED_CHARS)

        v = normalize(value)

        return false unless v.length == 9

        calculate_check_digit(v) == v[8].to_i
      end

      def format(value)
        v = normalize(value)

        return value unless v.length == 9

        "#{v[0...2]}.#{v[2...5]}.#{v[5...8]}-#{v[8...9]}"
      end
    end
  end
end

File: python/src/br_validator/ie/_am.py
"""AM: verified against the official SEFAZ-AM "Roteiro de Crítica da
Inscrição Estadual" (sintegra.gov.br/Cad_Estados/cad_AM.html). Format:
99.999.999-9 (8 digits + 1 check digit). Unlike most other states, when the
weighted sum itself is below 11 the digit is 11 minus the sum directly
(skipping the modulo step).
"""

import re

from .._shared import remove_non_digits
from ._mod11 import mod11_check_digit, weighted_sum

_ALLOWED_CHARS = re.compile(r"^[\d.\-\s]+$")
_WEIGHTS = [9, 8, 7, 6, 5, 4, 3, 2]


def normalize(value: str) -> str:
    return remove_non_digits(value)


def _calculate_check_digit(v: str) -> int:
    total = weighted_sum(v, _WEIGHTS)

    if total < 11:
        return 11 - total

    return mod11_check_digit(total)


def is_valid(value: str) -> bool:
    if not _ALLOWED_CHARS.match(value):
        return False

    v = normalize(value)

    if len(v) != 9:
        return False

    return _calculate_check_digit(v) == int(v[8])


def format(value: str) -> str:
    v = normalize(value)

    if len(v) != 9:
        return value

    return f"{v[:2]}.{v[2:5]}.{v[5:8]}-{v[8:9]}"
